File: services/dashboard/src/lib/volunteer/__tests__/pow.test.ts
// THE cross-language proof-of-work pin, mirroring services/infrastructure/pow/pow_test.go
// case-for-case. The digests, nonces, and leading-zero counts are hardcoded real outputs
// (independently reconfirmed with Node's crypto), so any divergence in the preimage
// layout, nonce encoding, or bit counting fails loudly here.

import { hexToBytes, leadingZeroBits, solvePow } from "../pow";
import { sha256 } from "../sha256";
import { bytesToHex } from "../identity";

// seqBytes returns n bytes whose values are start, start+1, … (mod 256) — the same
// one-line vector construction the Go golden test uses.
function seqBytes(start: number, n: number): Uint8Array {
  const b = new Uint8Array(n);
  for (let i = 0; i < n; i++) b[i] = (start + i) & 0xff;
  return b;
}

// manualDigest recomputes the solution digest independently of the solver, spelling out
// the documented preimage layout: SHA-256(challenge || publicKey || nonce as 8 big-endian
// bytes). This is the analogue of the Go test's manualDigest.
function manualDigest(
  challenge: Uint8Array,
  publicKey: Uint8Array,
  nonce: number
): Uint8Array {
  const nb = new Uint8Array(8);
  new DataView(nb.buffer).setUint32(4, nonce, false); // nonces here are < 2^32
  const preimage = new Uint8Array(challenge.length + publicKey.length + 8);
  preimage.set(challenge, 0);
  preimage.set(publicKey, challenge.length);
  preimage.set(nb, challenge.length + publicKey.length);
  return sha256(preimage);
}

// verifySolution mirrors the head's VerifySolution: a >= comparison of the digest's
// leading zero bits against the difficulty target.
function verifySolution(
  challenge: Uint8Array,
  publicKey: Uint8Array,
  nonce: number,
  difficultyBits: number
): boolean {
  return leadingZeroBits(manualDigest(challenge, publicKey, nonce)) >= difficultyBits;
}

describe("leadingZeroBits", () => {
  const highBitFirst = new Uint8Array(32);
  highBitFirst[0] = 0x80;
  const oneFirst = new Uint8Array(32);
  oneFirst[0] = 0x01;

  const cases: Array<[string, Uint8Array, number]> = [
    ["all-zero 32 bytes is fully saturated", new Uint8Array(32), 256],
    ["high bit set in first byte is zero", highBitFirst, 0],
    ["0x01 leaves seven leading zeros", oneFirst, 7],
    ["zero byte then 0xFF stops at eight", Uint8Array.from([0x00, 0xff]), 8],
    ["zero byte then 0x0F adds four", Uint8Array.from([0x00, 0x0f]), 12],
  ];

  it.each(cases)("%s", (_name, digest, want) => {
    expect(leadingZeroBits(digest)).toBe(want);
  });
});

describe("solvePow golden vector (cross-language pin with pow_test.go)", () => {
  const challenge = seqBytes(0, 32); // 0x00..0x1f
  const publicKey = seqBytes(32, 32); // 0x20..0x3f

  it("pins the nonce=1 preimage byte layout", () => {
    expect(bytesToHex(manualDigest(challenge, publicKey, 1))).toBe(
      "7f812700537ee9f8def5ab067d299b26f39ddf28875594ae0f95b6b1e40ce4c0"
    );
  });

  it("returns the first nonce (scanning from 0) that clears difficulty 16", () => {
    const sol = solvePow(challenge, publicKey, 16);
    expect(sol.nonce).toBe("19497");
    expect(sol.attempts).toBe(19498);
  });

  it("pins the difficulty-16 solution's digest and exact leading-zero count", () => {
    const d = manualDigest(challenge, publicKey, 19497);
    expect(bytesToHex(d)).toBe(
      "000040eb278da4930113209945a1b4d48614ba74bec414cc4b4418e1491e525b"
    );
    expect(leadingZeroBits(d)).toBe(17);
  });

  it("verifies as a >= comparison: accepted at 16 and 17, rejected at 18", () => {
    expect(verifySolution(challenge, publicKey, 19497, 16)).toBe(true);
    expect(verifySolution(challenge, publicKey, 19497, 17)).toBe(true);
    expect(verifySolution(challenge, publicKey, 19497, 18)).toBe(false);
  });
});

describe("solvePow public-key binding (diff 12)", () => {
  const challenge = seqBytes(0, 32); // 0x00..0x1f
  const keyA = seqBytes(32, 32); // 0x20..0x3f
  const keyB = seqBytes(64, 32); // 0x40..0x5f

  it("finds the pinned first nonce for key A", () => {
    expect(solvePow(challenge, keyA, 12).nonce).toBe("2286");
  });

  it("binds the public key into the preimage: A's nonce is not B's solution", () => {
    const dA = manualDigest(challenge, keyA, 2286);
    const dB = manualDigest(challenge, keyB, 2286);
    expect(bytesToHex(dA)).toBe(
      "000fb9757b2cac4a44c169521be41f7b7a7b56931dadb9d8bceab38e4a6b7e5c"
    );
    expect(bytesToHex(dB)).toBe(
      "9cc1515db812f8746ff8f4dc60d4f0240d47598b5473dc654373b2f2e51881cd"
    );
    expect(leadingZeroBits(dB)).toBe(0);
    expect(verifySolution(challenge, keyA, 2286, 12)).toBe(true);
    expect(verifySolution(challenge, keyB, 2286, 12)).toBe(false);
  });
});

describe("solvePow input validation", () => {
  it("throws when challenge is not 32 bytes", () => {
    expect(() => solvePow(new Uint8Array(31), new Uint8Array(32), 8)).toThrow();
  });

  it("throws when publicKey is not 32 bytes", () => {
    expect(() => solvePow(new Uint8Array(32), new Uint8Array(33), 8)).toThrow();
  });
});

describe("solvePow onProgress cadence", () => {
  const challenge = seqBytes(0, 32);
  const publicKey = seqBytes(32, 32);

  it("does not report progress when the solve finishes under one interval", () => {
    // The difficulty-16 golden solve takes 19,498 attempts, below the 65,536 cadence.
    const onProgress = jest.fn();
    solvePow(challenge, publicKey, 16, onProgress);
    expect(onProgress).not.toHaveBeenCalled();
  });

  it("reports progress at every 65,536th attempt for a longer solve", () => {
    // The first difficulty-18 solution is nonce 293300 (293,301 attempts), so progress
    // fires exactly at 65536, 131072, 196608, and 262144.
    const seen: number[] = [];
    const sol = solvePow(challenge, publicKey, 18, (attempts) => seen.push(attempts));
    expect(sol.nonce).toBe("293300");
    expect(seen).toEqual([65536, 131072, 196608, 262144]);
  });
});

describe("hexToBytes", () => {
  it("round-trips with bytesToHex", () => {
    const bytes = seqBytes(0, 32);
    expect(hexToBytes(bytesToHex(bytes))).toEqual(bytes);
  });

  it("decodes lowercase and uppercase hex", () => {
    expect(Array.from(hexToBytes("00ff10"))).toEqual([0x00, 0xff, 0x10]);
    expect(Array.from(hexToBytes("AABBCC"))).toEqual([0xaa, 0xbb, 0xcc]);
  });

  it("decodes the empty string to an empty array", () => {
    expect(hexToBytes("")).toEqual(new Uint8Array(0));
  });

  it("throws on odd length", () => {
    expect(() => hexToBytes("abc")).toThrow();
  });

  it("throws on non-hex characters", () => {
    expect(() => hexToBytes("zz")).toThrow();
    expect(() => hexToBytes("0x12")).toThrow();
  });
});

File: services/infrastructure/pow/pow_test.go
package pow

import (
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"testing"
)

// seqBytes returns n bytes whose values are start, start+1, … (mod 256). The golden
// vector below is built from three non-overlapping runs so a cross-language solver can
// reproduce the exact inputs from a one-line description.
func seqBytes(start, n int) []byte {
	b := make([]byte, n)
	for i := range b {
		b[i] = byte(start + i)
	}
	return b
}

// manualDigest recomputes the solution digest independently of this package's
// production code, spelling out the byte layout documented on VerifySolution:
//
//	SHA-256(challenge || publicKey || nonce as 8 big-endian bytes)
//
// The golden-vector test asserts this in-test recomputation equals the hardcoded hex
// constants, which is what actually pins the wire layout: if VerifySolution ever changed
// the preimage ordering or the nonce encoding, this helper and the constants would still
// agree but the real Solve / VerifySolution assertions would break.
func manualDigest(challenge, publicKey []byte, nonce uint64) []byte {
	var nb [8]byte
	binary.BigEndian.PutUint64(nb[:], nonce)
	h := sha256.New()
	h.Write(challenge)
	h.Write(publicKey)
	h.Write(nb[:])
	return h.Sum(nil)
}

// TestLeadingZeroBits pins the difficulty metric: the number of most-significant zero
// bits before the first set bit, counted across whole zero bytes and then within the
// first non-zero byte. This is the exact quantity difficulty targets are compared
// against, so a cross-language solver must count bits identically.
func TestLeadingZeroBits(t *testing.T) {
	cases := []struct {
		name   string
		digest []byte
		want   int
	}{
		{"all-zero 32 bytes is fully saturated", make([]byte, 32), 256},
		{"high bit set in first byte is zero", append([]byte{0x80}, make([]byte, 31)...), 0},
		{"0x01 leaves seven leading zeros", append([]byte{0x01}, make([]byte, 31)...), 7},
		{"zero byte then 0xFF stops at eight", []byte{0x00, 0xFF}, 8},
		{"zero byte then 0x0F adds four", []byte{0x00, 0x0F}, 12},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := LeadingZeroBits(tc.digest); got != tc.want {
				t.Errorf("LeadingZeroBits(%x) = %d, want %d", tc.digest, got, tc.want)
			}
		})
	}
}

// TestVerifySolution_GoldenVector is THE cross-implementation pin for the proof-of-work
// solution rule. The dashboard's TypeScript solver (and any future client) is validated
// against these exact bytes, so the values are hardcoded real outputs, not derived
// in-test. If any assertion here changes, every independent solver implementation must be
// updated in lockstep or it will stop producing solutions the head accepts.
//
// Vector definition (reproducible from this description alone):
//
//	challenge  = 32 bytes 0x00,0x01,…,0x1f   (byte i = i)
//	publicKey  = 32 bytes 0x20,0x21,…,0x3f   (byte i = 32+i)
//	digest(n)  = SHA-256(challenge || publicKey || n as 8 big-endian bytes)
func TestVerifySolution_GoldenVector(t *testing.T) {
	const (
		// digest for nonce = 1. Pins the preimage byte layout (order + big-endian
		// nonce encoding). Its own leading-zero count is only 1, so it is not a
		// solution at any real difficulty — it exists purely to fix the layout.
		nonce1DigestHex = "7f812700537ee9f8def5ab067d299b26f39ddf28875594ae0f95b6b1e40ce4c0"

		// The first nonce (scanning from 0) whose digest clears difficulty 16. It
		// happens to clear 17 bits, so difficulty 17 also accepts it and difficulty
		// 18 rejects it — pinned exactly below.
		diff16Nonce     = uint64(19497)
		diff16DigestHex = "000040eb278da4930113209945a1b4d48614ba74bec414cc4b4418e1491e525b"
		diff16LZB       = 17
	)

	challenge := seqBytes(0, 32)  // 0x00..0x1f
	publicKey := seqBytes(32, 32) // 0x20..0x3f

	// (1) Byte-layout pin: an in-test SHA-256 over the documented preimage must equal
	// the hardcoded nonce=1 digest.
	if got := hex.EncodeToString(manualDigest(challenge, publicKey, 1)); got != nonce1DigestHex {
		t.Fatalf("nonce=1 digest = %s, want %s (preimage byte layout changed)", got, nonce1DigestHex)
	}

	// (2) The reference solver must return the pinned difficulty-16 nonce as the FIRST
	// solution. This is the strongest cross-language assertion: any solver scanning
	// nonces from 0 must arrive at exactly this value.
	if got := Solve(challenge, publicKey, 16); got != diff16Nonce {
		t.Fatalf("Solve(diff=16) = %d, want %d", got, diff16Nonce)
	}

	// (3) The difficulty-16 solution's digest is fixed, and so is its exact leading-zero
	// count.
	d16 := manualDigest(challenge, publicKey, diff16Nonce)
	if got := hex.EncodeToString(d16); got != diff16DigestHex {
		t.Fatalf("diff-16 digest = %s, want %s", got, diff16DigestHex)
	}
	if got := LeadingZeroBits(d16); got != diff16LZB {
		t.Fatalf("LeadingZeroBits(diff-16 digest) = %d, want %d", got, diff16LZB)
	}

	// (4) Verification is a >= comparison against the difficulty target: the nonce is
	// accepted at every difficulty up to its actual leading-zero count and rejected
	// above it.
	if !VerifySolution(challenge, publicKey, diff16Nonce, 16) {
		t.Errorf("VerifySolution(diff=16) = false, want true")
	}
	if !VerifySolution(challenge, publicKey, diff16Nonce, diff16LZB) {
		t.Errorf("VerifySolution(diff=%d) = false, want true", diff16LZB)
	}
	if VerifySolution(challenge, publicKey, diff16Nonce, diff16LZB+1) {
		t.Errorf("VerifySolution(diff=%d) = true, want false", diff16LZB+1)
	}
}

// TestSolve_RoundTrip checks the reference solver against verification on an input
// distinct from the golden vector: Solve must return a nonce that VerifySolution accepts
// at the same difficulty. Difficulty 8 keeps the brute force fast (~256 attempts).
func TestSolve_RoundTrip(t *testing.T) {
	const difficulty = 8

	// A fixed but non-sequential input, to show the round trip holds generally and not
	// only for the golden byte runs.
	challenge := make([]byte, 32)
	publicKey := make([]byte, 32)
	for i := range challenge {
		challenge[i] = byte(i*7 + 1)
		publicKey[i] = byte(i*13 + 5)
	}

	nonce := Solve(challenge, publicKey, difficulty)
	if !VerifySolution(challenge, publicKey, nonce, difficulty) {
		t.Fatalf("VerifySolution(Solve() = %d) = false, want true at difficulty %d", nonce, difficulty)
	}
	// The returned nonce must genuinely clear the target, not merely be accepted by a
	// buggy comparison.
	if got := LeadingZeroBits(manualDigest(challenge, publicKey, nonce)); got < difficulty {
		t.Errorf("solution digest has %d leading zero bits, want >= %d", got, difficulty)
	}
}

// TestVerifySolution_PubkeyBinding pins that the public key is part of the solution
// preimage: a nonce solved for key A is not a solution for a different key B. This is why
// a redeemed solution is single-purpose (see internal/admission's RedeemChallenge). The
// values are hardcoded from the reference solver so the negative case is deterministic —
// B's digest here has zero leading zero bits, well under the difficulty, so there is no
// probabilistic flake.
func TestVerifySolution_PubkeyBinding(t *testing.T) {
	const (
		difficulty = 12
		// First nonce (from 0) solving difficulty 12 for key A.
		bindNonce = uint64(2286)
		// A's digest at bindNonce clears exactly 12 bits; B's clears 0.
		aDigestHex = "000fb9757b2cac4a44c169521be41f7b7a7b56931dadb9d8bceab38e4a6b7e5c"
		bDigestHex = "9cc1515db812f8746ff8f4dc60d4f0240d47598b5473dc654373b2f2e51881cd"
	)

	challenge := seqBytes(0, 32) // 0x00..0x1f
	keyA := seqBytes(32, 32)     // 0x20..0x3f
	keyB := seqBytes(64, 32)     // 0x40..0x5f

	// The reference solver finds the pinned nonce for A.
	if got := Solve(challenge, keyA, difficulty); got != bindNonce {
		t.Fatalf("Solve(keyA, diff=%d) = %d, want %d", difficulty, got, bindNonce)
	}

	// Same challenge and nonce, different key ⇒ different digest. Pin both.
	dA := manualDigest(challenge, keyA, bindNonce)
	dB := manualDigest(challenge, keyB, bindNonce)
	if got := hex.EncodeToString(dA); got != aDigestHex {
		t.Fatalf("keyA digest = %s, want %s", got, aDigestHex)
	}
	if got := hex.EncodeToString(dB); got != bDigestHex {
		t.Fatalf("keyB digest = %s, want %s", got, bDigestHex)
	}
	if hex.EncodeToString(dA) == hex.EncodeToString(dB) {
		t.Fatalf("keyA and keyB produced the same digest; public key is not bound into the preimage")
	}
	if got := LeadingZeroBits(dB); got != 0 {
		t.Fatalf("keyB digest has %d leading zero bits, want 0 (deterministic-failure precondition)", got)
	}

	// The solution is valid for A and, deterministically, invalid for B.
	if !VerifySolution(challenge, keyA, bindNonce, difficulty) {
		t.Errorf("VerifySolution(keyA) = false, want true")
	}
	if VerifySolution(challenge, keyB, bindNonce, difficulty) {
		t.Errorf("VerifySolution(keyB) = true, want false (nonce must not carry across keys)")
	}
}
